src/libraries/Microsoft.PowerFx.Core/Syntax/NamedFormula.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Syntax.Nodes;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Syntax
+{
+    /// <summary>
+    /// This encapsulates a named formula: its original script, the parsed result, and any parse errors.
+    /// </summary>
+    internal class NamedFormula
+    {
+        /// <summary>
+        /// A script containing one or more named formulas.
+        /// </summary>
+        public readonly string Script;
+
+        // The language settings used for parsing this script.
+        // May be null if the script is to be parsed in the current locale.
+        public readonly ILanguageSettings Loc;
+
+        public Dictionary<DName, TexlNode> FormulasResult;
+
+        public bool IsParsed => FormulasResult != null;
+
+        public bool HasParseErrors { get; private set; }
+
+        private List<TexlError> _errors;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NamedFormula"/> class.
+        /// </summary>
+        /// <param name="script"></param>
+        /// <param name="loc"></param>
+        public NamedFormula(string script, ILanguageSettings loc = null)
+        {
+            Contracts.AssertValue(script);
+            Contracts.AssertValueOrNull(loc);
+
+            Script = script;
+            Loc = loc;
+        }
+
+        /// <summary>
+        /// Ensures that the named formulas have been parsed and if not, parses them.
+        /// </summary>
+        /// <returns></returns>
+        public bool EnsureParsed()
+        {
+            if (FormulasResult == null)
+            {
+                Contracts.AssertValue(Script);
+                Contracts.AssertValueOrNull(Loc);
+                var result = TexlParser.ParseFormulasScript(Script, loc: Loc);
+                FormulasResult = result.NamedFormulas;
+                _errors = result.Errors;
+                HasParseErrors = result.HasError;
+                Contracts.AssertValue(FormulasResult);
+            }
+
+            return _errors == null;
+        }
+
+        /// <summary>
+        /// Returns the formula portion of the script for a given named formula identifier.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="subScript"></param>
+        /// <returns></returns>
+        public bool TryGetSubscript(DName name, out string subScript)
```

Comment:
>string [](http://example.com/codeflow?start=52&length=6)

curious: who needs the original string (as opposed to the TexlNode)

Reply:
Most of PowerApps-Client expects a rule to have a string as well as a parse tree. It's mostly used for telemetry, but is in a large enough part of the codebase that it's easier to meet that assumption

Reply:
Consider returning Formula (or even just enumerating formulas).